### PR TITLE
subnet-grouping2

### PR DESCRIPTION
**Pull request by haim-kermany**
git did not agree with the previews one.
Adi, please review the vpcmodel\ibmvpc parts

### CODE REVIEW

**Review by zivnevo — CHANGES_REQUESTED** [review]
Some comments before I start looking into `layout.go` and `subnetsLayout.go`

File: pkg/vpcmodel/drawioOutput.go [hunk 1]
```diff
@@ -119,11 +141,19 @@ func (d *DrawioOutputFormatter) WriteOutput(c1, c2 *VPCConfig,
...
-	case AllSubnets, SingleSubnet:
-		err = errors.New("SubnetLevel/SingleSubnet use case not supported for draw.io format")
+		var gConn *GroupConnLines
+		if conn != nil {
+			gConn = conn.GroupedConnectivity
+		}
+		return d.writeOutputGeneric(c1, gConn, outFile, false)
+	case AllSubnets:
+		var gConn *GroupConnLines
+		if subnetsConn != nil {
```
Comment (ShiriMoran) [inline pkg/vpcmodel/drawioOutput.go:151]: Why is this check, of comparing to nil, required here? It is not performed in the other formats.

**Review by ShiriMoran — COMMENTED** [review]
@haim-kermany I've put some initial comments. Will still need to take it off line to get some basic understanding.

File: pkg/drawio/squareTreeNode.go [hunk 2]
```diff
@@ -264,3 +271,46 @@ func (tn *GroupSquareTreeNode) setVisibility(visibility groupSquareVisibility) {
...
 	return nil, append(tn.elements, tn.groupedIcons...), tn.connections
 }
+
+// ////////////////////////////////////////////////////////////////////////////
+type GroupSubnetsSquareTreeNode struct {
+	abstractSquareTreeNode
+	groupedSubnets []SquareTreeNodeInterface
+}
+
+func GroupedSubnetsSquare(parent *VpcTreeNode, groupedSubnets []SquareTreeNodeInterface) SquareTreeNodeInterface {
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/squareTreeNode.go:281]: in cases the TN is not always generated, the name is not New*TreeNode.
in this case, GroupedSubnetsSquare() might return an existing square
please notice that GroupedSubnetsSquare() calles newGroupSubnetsSquareTreeNode()

File: .github/workflows/golangci-lint.yml [hunk 3]
```diff
@@ -13,6 +13,6 @@ jobs:
     steps:
       - uses: actions/setup-go@93397bea11091df50f3d7e59dc26a7711a8bcfbe
         with:
-          go-version: 1.18
+          go-version: "1.20"
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply .github/workflows/golangci-lint.yml]: fixed

File: .github/workflows/go-test.yml [hunk 4]
```diff
@@ -18,7 +18,7 @@ jobs:
     - name: Set up Go
       uses: actions/setup-go@93397bea11091df50f3d7e59dc26a7711a8bcfbe
       with:
-        go-version: 1.18
+        go-version: "1.20"
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply .github/workflows/go-test.yml]: fixed

File: pkg/vpcmodel/drawioOutput.go [hunk 5]
```diff
@@ -50,30 +63,37 @@ func (d *DrawioOutputFormatter) createDrawioTree() {
...
+		if !d.subnetMode || ns.ShowOnSubnetMode() {
+			d.gen.TreeNode(ns)
+		}
 	}
 }
 
 func (d *DrawioOutputFormatter) createNodes() {
 	for _, n := range d.cConfig.Nodes {
-		if !n.IsExternal() {
+		if !n.IsExternal() && (!d.subnetMode || n.ShowOnSubnetMode()) {
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/vpcmodel/drawioOutput.go]: they are created on the fly, during the creation of the edges

File: pkg/vpcmodel/drawioGenerator.go [hunk 6]
```diff
@@ -8,6 +8,7 @@ import (
 type DrawioResourceIntf interface {
 	GenerateDrawioTreeNode(gen *DrawioGenerator) drawio.TreeNodeInterface
 	IsExternal() bool
+	ShowOnSubnetMode() bool
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/vpcmodel/drawioGenerator.go:11]: some return false, see ibmDrawioGenerator.go

File: pkg/vpcmodel/drawioOutput.go [hunk 7]
```diff
@@ -119,11 +141,19 @@ func (d *DrawioOutputFormatter) WriteOutput(c1, c2 *VPCConfig,
...
+		}
+		return d.writeOutputGeneric(c1, gConn, outFile, false)
+	case AllSubnets:
+		var gConn *GroupConnLines
+		if subnetsConn != nil {
+			gConn = subnetsConn.GroupedConnectivity
+		}
+		return d.writeOutputGeneric(subnetsConn.VPCConfig, gConn, outFile, true)
+	case SingleSubnet:
+		return &SingleAnalysisOutput{}, errors.New("DebugSubnet use case not supported for draw.io format currently ")
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/vpcmodel/drawioOutput.go]: fixed

File: pkg/vpcmodel/drawioOutput.go [hunk 8]
```diff
@@ -50,30 +63,37 @@ func (d *DrawioOutputFormatter) createDrawioTree() {
 
 func (d *DrawioOutputFormatter) createNodeSets() {
 	for _, ns := range d.cConfig.NodeSets {
-		d.gen.TreeNode(ns)
+		if !d.subnetMode || ns.ShowOnSubnetMode() {
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/vpcmodel/drawioOutput.go]: done

File: pkg/vpcmodel/drawioOutput.go [hunk 9]
```diff
@@ -25,23 +25,36 @@ func (e *edgeInfo) IsExternal() bool {
...
+func (d *DrawioOutputFormatter) init(cConfig *VPCConfig, conn *GroupConnLines, subnetMode bool) {
 	d.cConfig = cConfig
 	d.conn = conn
+	d.subnetMode = subnetMode
 	d.gen = NewDrawioGenerator(cConfig.CloudName)
+	d.gen.network.SubnetMode = subnetMode
 	d.routers = map[drawio.TreeNodeInterface]drawio.IconTreeNodeInterface{}
 }
 
+func (d *DrawioOutputFormatter) writeOutputGeneric(cConfig *VPCConfig, conn *GroupConnLines, outFile string, subnetMode bool) (
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/vpcmodel/drawioOutput.go]: done

File: pkg/drawio/lineTreeNode.go [hunk 10]
```diff
@@ -86,29 +86,24 @@ func NewConnectivityLineTreeNode(network SquareTreeNodeInterface,
...
-		iconDst = NewGroupPointTreeNode(dst.(SquareTreeNodeInterface), directed, false, "")
-	} else {
-		iconDst = dst.(IconTreeNodeInterface)
+	if dst.IsSquare() && dst.(SquareTreeNodeInterface).IsGroupingSquare() {
+		dst = NewGroupPointTreeNode(dst.(SquareTreeNodeInterface), directed, false, "")
 	}
-	if iconSrc.IsGroupingPoint() {
-		iconSrc.(*GroupPointTreeNode).setColleague(iconDst)
+	if src.IsIcon() && src.(IconTreeNodeInterface).IsGroupingPoint() {
+		src.(*GroupPointTreeNode).setColleague(dst.(IconTreeNodeInterface))
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/lineTreeNode.go:96]: its guaranteed that GroupingPoints are connected to icons

File: pkg/drawio/squareTreeNode.go [hunk 11]
```diff
@@ -60,6 +64,7 @@ type NetworkTreeNode struct {
 	abstractSquareTreeNode
 	clouds        []SquareTreeNodeInterface
 	publicNetwork SquareTreeNodeInterface
+	SubnetMode    bool
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/squareTreeNode.go]: fixed

File: pkg/vpcmodel/drawioOutput.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/vpcmodel/drawioOutput.go]: fixed

File: pkg/drawio/createMapFile.go [hunk 12]
```diff
@@ -44,11 +44,15 @@ func (data *drawioData) ElementComment(tn TreeNodeInterface) string {
 // 2. we put the icons above the squares so we can mouse over it for tooltips
 // 3. we put the sgs in the bottom. if a sg is above a square, it will block the the tooltip of the children of the square.
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/createMapFile.go]: pushed

File: pkg/drawio/layout.go [hunk 13]
```diff
@@ -288,6 +298,100 @@ func (ly *layoutS) layoutSubnetsIcons() {
...
+						ly.setDefaultLocation(subnet, rowIndex, zonesCol[zone])
+						rowIndex++
+					}
+				}
+			}
+		}
+	}
+}
+
+func (ly *layoutS) resolveGroupedSubnetsOverride() {
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/layout.go]: fixed

File: pkg/drawio/subnetsLayout.go [hunk 14]
```diff
@@ -0,0 +1,826 @@
...
+func (s *genericSet[T]) isIntersect(s2 *genericSet[T]) bool {
+	for i := range *s {
+		if (*s2)[i] {
+			return true
+		}
+	}
+	return false
+}
+func (s *genericSet[T]) copy() genericSet[T] {
+	c := genericSet[T]{}
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/subnetsLayout.go]: done

File: pkg/drawio/subnetsLayout.go
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/subnetsLayout.go:1]: moved to common package

File: pkg/drawio/subnetsLayout.go [hunk 15]
```diff
@@ -0,0 +1,770 @@
...
+// todo: how to remove???
+func (s *groupTnSet) AsKey() setAsKey {
+	return setAsKey(((*common.GenericSet[TreeNodeInterface])(s)).AsKey())
+}
+func (s *groupSet) AsKey() setAsKey { return setAsKey(((*common.GenericSet[*groupDataS])(s)).AsKey()) }
+func (s *miniGroupSet) IsIntersect(s2 *miniGroupSet) bool {
+	return ((*common.GenericSet[*miniGroupDataS])(s)).IsIntersect((*common.GenericSet[*miniGroupDataS])(s2))
+}
+func (s *groupSet) AsList() []*groupDataS { return ((*common.GenericSet[*groupDataS])(s)).AsList() }
+
```
Comment (zivnevo) [inline pkg/drawio/subnetsLayout.go:48]:
```suggestion
type subnetSet = common.GenericSet[TreeNodeInterface]
type groupTnSet = common.GenericSet[TreeNodeInterface]
type groupSet = common.GenericSet[*groupDataS]
type miniGroupSet = common.GenericSet[*miniGroupDataS]
type setAsKey = common.SetAsKey
```

**Review by zivnevo — CHANGES_REQUESTED** [review]
A few more

File: pkg/common/genericSet.go [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+	ss := []string{}
+	for i := range *s {
+		if _, ok := interfaceIndex[i]; !ok {
+			interfaceIndex[i] = len(interfaceIndex)
+		}
+		ss = append(ss, strconv.Itoa(interfaceIndex[i]))
+	}
+	sort.Strings(ss)
+	return SetAsKey(strings.Join(ss, ","))
+}
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/common/genericSet.go:31]: OK
minor comment:
pointers that are not garbage-collected, are not reused

File: pkg/vpcmodel/drawioOutput.go [hunk 17]
```diff
@@ -25,23 +25,36 @@ func (e *edgeInfo) IsExternal() bool {
...
 
 type DrawioOutputFormatter struct {
-	cConfig *VPCConfig
-	conn    *VPCConnectivity
-	gen     *DrawioGenerator
-	routers map[drawio.TreeNodeInterface]drawio.IconTreeNodeInterface
+	cConfig    *VPCConfig
+	conn       *GroupConnLines
+	gen        *DrawioGenerator
+	subnetMode bool
```
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/vpcmodel/drawioOutput.go]: indeed

File: pkg/vpcmodel/drawioGenerator.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ShiriMoran) [reply pkg/vpcmodel/drawioGenerator.go:11]: right

File: pkg/common/genericSet.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/common/genericSet.go]: fixed, 
I removed interfaceIndex.
instead, I use sprintf to convert a pointer to a string

File: pkg/drawio/subnetsLayout.go [hunk 18]
```diff
@@ -0,0 +1,770 @@
...
+type groupDataS struct {
+	miniGroups    miniGroupSet
+	subnets       subnetSet
+	treeNode      TreeNodeInterface
+	children      groupSet
+	toSplitGroups groupSet
+	splitFrom     groupSet
+	splitTo       groupSet
+}
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/drawio/subnetsLayout.go]: done

File: pkg/common/genericSet.go [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+// the main functionality of genericSet is asKey() - conversion to a string.
+// asKey() is needed for using genericSet as a key of a map
+// //////////////////////////////////////////////////////////////////////////////////////////////
+
+type GenericSet[T comparable] map[T]bool
+type SetAsKey string
+
+var interfaceIndex map[interface{}]int = map[interface{}]int{}
+
+func (s *GenericSet[T]) AsKey() SetAsKey {
```
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/common/genericSet.go]: done

File: pkg/drawio/subnetsLayout.go (hunk 15 above)
Comment (haim-kermany) [reply pkg/drawio/subnetsLayout.go]: done
works, Tnx

File: pkg/common/genericSet.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (haim-kermany) [reply pkg/common/genericSet.go]: fixed, using reflact() to get the pointer

**Review by zivnevo — APPROVED** [review]
ok to merge

**Pull request merged**